routing/router.go

```diff
@@ -1694,9 +1704,67 @@ func (r *ChannelRouter) sendToSwitch(route *route.Route, paymentHash [32]byte) (
 	firstHop := lnwire.NewShortChanIDFromInt(
 		route.Hops[0].ChannelID,
 	)
-	return r.cfg.SendToSwitch(
-		firstHop, htlcAdd, circuit,
+
+	// Using the created circuit, initialize the error
+	// decrypter so we can parse+decode any failures
+	// incurred by this payment within the switch.
+	errorDecryptor := &htlcswitch.SphinxErrorDecrypter{
+		OnionErrorDecrypter: sphinx.NewOnionErrorDecrypter(circuit),
+	}
+
+	// We generate a new, unique payment ID that we will use for
+	// this HTLC.
+	paymentID, err := r.cfg.NextPaymentID()
+	if err != nil {
+		return [32]byte{}, true, err
+	}
+
+	err = r.cfg.Payer.SendHTLC(
+		firstHop, paymentID, htlcAdd,
+	)
+	if err != nil {
+		log.Errorf("Failed sending attempt %d for payment %x to "+
+			"switch: %v", paymentID, paymentHash, err)
+
+		// We must inspect the error to know whether it was critical or
+		// not, to decide whether we should continue trying.
+		finalOutcome := r.processSendError(
```

Comment:
This is not the type of error that is suitable for `processSendError`. It will never be a wire error at this stage and `finalOutcome` is always `true`?

Reply:
Yeah I think this will be the case, as if we get an error here, then it never left the switch. Will need to double check before we modify this at all. But in the end, having this statement here may actually be harmless.

Reply:
We will return a `ForwardingError` for some local cases, like if no Link has capacity.

Reply:
Hmm I see. That is a whole new chapter in failure message interpretation then. I am wondering if it would be better in the end to gradually move those over to local errors that describe the exact problem, instead of a text string `ExtraMsg`.